util/fipstools/inject_hash/macho_parser/macho_parser.c

```diff
@@ -0,0 +1,165 @@
+            segment_load_cmd *segment = (segment_load_cmd *)&load_commands[i];
+            if (strcmp(segment->segname, "__TEXT") == 0) {
+                section_data *sections = (section_data *)&segment[1];
+                for (uint32_t j = 0; j < segment->nsects; j++) {
+                    if (strcmp(sections[j].sectname, "__text") == 0 || strcmp(sections[j].sectname, "__const") == 0) {
+                        macho->sections[section_index].offset = sections[j].offset;
+                        macho->sections[section_index].size = sections[j].size;
+                        strcpy(macho->sections[section_index].name, sections[j].sectname);
+                        section_index++;
+                    }
+                }
+            }
+        } else if (load_commands[i].cmd == LC_SYMTAB) {
+            symtab_load_cmd *symtab = (symtab_load_cmd *)&load_commands[i];
+            macho->sections[section_index].offset = symtab->symoff;
+            macho->sections[section_index].size = symtab->nsyms * sizeof(symbol_info);
+            strcpy(macho->sections[section_index].name, "__symbol_table");
+            section_index++;
+            macho->sections[section_index].offset = symtab->stroff;
+            macho->sections[section_index].size = symtab->strsize;
+            strcpy(macho->sections[section_index].name, "__string_table");
+            section_index++;
+        }
+    }
+
+
+end:
+    if (load_commands != NULL) {
+        free(load_commands);
+    }
+    if (file != NULL) {
+        fclose(file);
+    }
+    return ret;
+}
+
+void free_macho_file(machofile *macho) {
+    free(macho->sections);
+    free(macho);
+    macho = NULL;
+}
+
+uint8_t* get_macho_section_data(const char *filename, machofile *macho, const char *section_name, size_t *size, uint32_t *offset) {
+    FILE *file = NULL;
+    uint8_t *section_data = NULL;
+    uint8_t *ret = NULL;
+
+    file = fopen(filename, "rb");
+    if (!file) {
+        LOG_ERROR("Error opening file %s", filename);
+        goto end;
+    }
+    for (uint32_t i = 0; i < macho->num_sections; i++) {
+        if (strcmp(macho->sections[i].name, section_name) == 0) {
+            section_data = malloc(macho->sections[i].size);
+            if (!section_data) {
+                LOG_ERROR("Error allocating memory for section data");
+                goto end;
+            }
+
+            fseek(file, macho->sections[i].offset, SEEK_SET);
+            fread(section_data, 1, macho->sections[i].size, file);
+
+            if (size != NULL) {
+                *size = macho->sections[i].size;
+            }
+            if (offset) {
+                *offset = macho->sections[i].offset;
+            }
+
+            ret = section_data;
+            goto end;
+        }
+    }
+
+    LOG_ERROR("Section %s not found in macho file %s", section_name, filename);
+end:
+    if (file != NULL) {
+        fclose(file);
+    }
+    if (ret == NULL && section_data != NULL) {
```

Comment:
This condition doesn't appear to be possible.  (Moving the `section_data` declaration into the for loop should provide more clarity on this.)